Date: Wed, 12 May 2021 13:26:18 +1000
Subject: x86: Disable KASLR when Xen is detected

There's currently an issue with Xen and KASLR causing hibernation to
break (and possibly kexec/kdump too). Until we have got to the bottom
of this and fixed the root cause, let's disable KASLR at runtime when
running on Xen instances so we can enable it for Nitro.

This also adds a boot message to match ARM and help detect whether
this test worked as expected.

Signed-off-by: Benjamin Herrenschmidt <benh@amazon.com>
(cherry picked from commit 53eb92c81d710e9da9df2ed76572488bc2265769)
Signed-off-by: Maximilian Heyne <mheyne@amazon.de>
[v6.18: contextual conflict with asm/setup.h]
Signed-off-by: Shaoying Xu <shaoyi@amazon.com>

diff --git a/arch/x86/boot/compressed/kaslr.c b/arch/x86/boot/compressed/kaslr.c
--- a/arch/x86/boot/compressed/kaslr.c
+++ b/arch/x86/boot/compressed/kaslr.c
@@ -1,3 +1,4 @@
+
 // SPDX-License-Identifier: GPL-2.0
 /*
  * kaslr.c
@@ -28,6 +29,9 @@
 #include <generated/utsversion.h>
 #include <generated/utsrelease.h>
 
+/* xen_cpuid_base/hypervisor_cpuid_base inlines */
+#include <asm/xen/hypervisor.h>
+
 #define _SETUP
 #include <asm/setup.h>	/* For COMMAND_LINE_SIZE */
 #undef _SETUP
@@ -870,6 +874,10 @@ void choose_random_location(unsigned long input,
 		warn("KASLR disabled: 'nokaslr' on cmdline.");
 		return;
 	}
+	if (xen_cpuid_base() != 0) {
+		warn("KASLR disabled: Xen hypervisor detected.");
+		return;
+	}
 
 	boot_params_ptr->hdr.loadflags |= KASLR_FLAG;
 
diff --git a/arch/x86/kernel/setup.c b/arch/x86/kernel/setup.c
--- a/arch/x86/kernel/setup.c
+++ b/arch/x86/kernel/setup.c
@@ -56,6 +56,7 @@
 #include <asm/thermal.h>
 #include <asm/unwind.h>
 #include <asm/vsyscall.h>
+#include <asm/setup.h>
 
 /*
  * max_low_pfn_mapped: highest directly mapped pfn < 4 GB
@@ -905,6 +906,9 @@ void __init setup_arch(char **cmdline_p)
 	printk(KERN_INFO "Command line: %s\n", boot_command_line);
 	boot_cpu_data.x86_phys_bits = MAX_PHYSMEM_BITS;
 #endif
+#ifdef CONFIG_RANDOMIZE_BASE
+	printk(KERN_INFO "KASLR %s\n", kaslr_enabled() ? "enabled" : "disabled");
+#endif
 
 #ifdef CONFIG_CMDLINE_BOOL
 #ifdef CONFIG_CMDLINE_OVERRIDE
